Add fallback to URL encoding to account for special chars

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #8791

## Description
This PR addresses a crash found in TestFlight for 12.1 during beta testing

When unwrapping URLs from `WooConstants` we trust that these are well-formed and do not contain any un-escaped character, we perform this check [here](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/System/WooConstants.swift#L272-L278) or return a `fatalError` otherwise.

This PR adds a fallback to attempt and escape the URL format before returning this error

## Testing instructions
Confirm that Unit Tests pass

iamgabrielma committed:
Add fallback to URL encoding to account for special chars

itsmeichigo commented on WooCommerce/Classes/System/WooConstants.swift:
```diff
@@ -162,9 +162,9 @@ extension WooConstants {
-        case inPersonPaymentsCashOnDeliveryFeedback,
-             inPersonPaymentsFirstTransactionFeedback,
-             inPersonPaymentsPowerUsersFeedback = "https://automattic.survey.fm/woo-app-ipp-in-app-feedback-testing"
+        case inPersonPaymentsCashOnDeliveryFeedback = "https://automattic.survey.fm/woo-app-–-cod-survey"
+        case inPersonPaymentsFirstTransactionFeedback = "https://automattic.survey.fm/woo-app-–-ipp-first-transaction-survey"
+        case inPersonPaymentsPowerUsersFeedback = "https://automattic.survey.fm/woo-app-–-ipp-survey-for-power-users"
```
Super nit: you may not need to update URLs for PR for testing purposes - instead adding an instruction in the testing steps should be enough.

iamgabrielma replied:
Yes, that makes sense 🙇  I'll leave these for the moment though just for safety, as initially the `#if/else DEBUG` flag didn't jump on those URLs having un-escaped characters when I was testing. I added this bit to the project i2, along with more extensive testing for other potential un-escaped characters.

itsmeichigo commented on WooCommerce/Classes/System/WooConstants.swift:
```diff
@@ -270,10 +270,10 @@ private extension WooConstants.URLs {
+            let fallback = URL(string: url.addingPercentEncoding(withAllowedCharacters: NSCharacterSet.urlQueryAllowed) ?? "")
+            return fallback ?? URL(string: "")!
         }
+        return url
```
Super nit: Since we want to crash this method when the URL cannot be parsed, this can be updated for readability:
```suggestion
        if let url = URL(string: url) {
            return URL
        } else if let escapedString = url.addingPercentEncoding(withAllowedCharacters: NSCharacterSet.urlQueryAllowed),
            let escapedURL = URL(string: escapedString) {
                return escapedURL
        } else {
            fatalError("Expected URL \(url) to be a well-formed URL.")
        }
```

iamgabrielma replied:
Good point, changed here: https://github.com/woocommerce/woocommerce-ios/pull/8805/commits/a68ef434cefc58085d059155ecdff73aa2366751

As a side note I'm not positive that we should crash if URLs are un-escaped in this way from an UX perspective, or just redirect users to a different WooCommerce URL, but that's something I'll explore on a different PR 👍

itsmeichigo commented on WooCommerce/WooCommerceTests/System/WooConstantsTests.swift:
```diff
@@ -12,7 +12,32 @@ final class WooConstantsTests: XCTestCase {
+        let urlsWithSpecialCharacters = [
+            WooConstants.URLs.inPersonPaymentsCashOnDeliveryFeedback,
+            WooConstants.URLs.inPersonPaymentsFirstTransactionFeedback,
+            WooConstants.URLs.inPersonPaymentsPowerUsersFeedback
```
Nit: we should use test string instead of `WooConstants`, in case the URLs are changed. Also, we should use test URLs instead of actual URLs in case there are privacy issues.
```suggestion
            "https://test.com/woo-app-–-cod-survey",
```

iamgabrielma replied:
Definitely, changed here: https://github.com/woocommerce/woocommerce-ios/pull/8805/commits/0e4a799709ff01bbc48785e1adfbbfa8dd5f8bc2

iamgabrielma replied:
As an update, I had to add a testing string to `WooConstants.URL` in order to test these cases. The reason is because `asURL()` is a private extension, the one we get directly comes from AlamoFire and would crash if we pass a non-trusted URL. Our version seems was a re-implementation of this one.

```
▿ AFError
  ▿ invalidURL : 1 element
    - url : "https://test.com/test-–-survey"*/
``` 

Change on fc7dc937c68c36ab8bd1ca2f23e375adf8051c6e

itsmeichigo replied:
I didn't know `asURL()` is an internal method of `WooConstants` - in that case maybe it's better to add an extension method for `URL` and write tests for that extension instead.

Otherwise, you should add that testURL to an extension of `WooConstants` within the Test target - that looks better than adding it to the project @iamgabrielma.

iamgabrielma replied:
> I didn't know asURL() is an internal method of WooConstants - in that case maybe it's better to add an extension method for URL and write tests for that extension instead.

Sounds better indeed. Opened a new issue at https://github.com/woocommerce/woocommerce-ios/issues/8810 to deal with this separately

itsmeichigo reviewed (approved):
This works as expected but I have quite a few nits 😅 Please feel free to merge after updating.

iamgabrielma committed:
Update trustedURL methot to crash if needed

iamgabrielma committed:
Update test with un-escaped string rather than WooConstant reference

iamgabrielma committed:
Update WooConstants with an un-scaped testing URL

`asURL()` is a private extension located on WooConstants, if we use it directle over a testing URL in the Unit Test, then the AlamoFire one will be called instead, which doesn’t deal with un-scaped strings and would keep crashing.

malinajirka commented:
We realized these changes were merged directly into the release branch - `release/12.1`. Since updating `trunk` with the release branch is part of the release process this PR isn't necessary. 

Either case, thanks for the fixes @iamgabrielma and thanks for the review @itsmeichigo!
